source/cluster_manager/clusterManager.js

```diff
@@ -555,204 +566,945 @@ var ClusterManager = function (clusterName, selfId, spec) {
+    role: "follower",
+    timestamp: (new Date()).getTime()
+}
+
+var leaderMsgHandler = null;
+
+var runAsFollower = function (topicChannel, manager) {
+    var isInstallSnapshot = false;
+    var verifyLeaderCallback = [];
+    var isVerifyLeader = false;
+    var verifyLeaderResult = false;
+    state.role = "follower";
+    state.timestamp=(new Date()).getTime();
+
+    var routerKeys = [`clusterManager.follower.${state.id}`];
+    var observer = new HeartbeatObserver(async (loseTime,lastContact)=>{
+        let begin = (new Date()).getTime();
+        await mutex.lock("follower.HeartbeatObserver");
+        try{
+            if(state.role !== "follower"){
+                log.warn('follower cycle in HeartbeatObserver',"self:", state);
+                return;
+            }
+
+            //check if leader is still alive
+            if (false && state.leaderId != 0) {
+                let count = 2;
+                while (count > 0) {
+                    count--;
+                    if (await verifyLeader()) {
+                        observer.resetTimer();
+                        return;
+                    }
+                }
+            }
+            log.fatal("Lose heart-beat from leader:",state.leaderId,"loseTime:", loseTime, 'lastContact:', lastContact, "timeout:",`${loseTime.getTime()-lastContact.getTime()}ms`);
+            await changeRole("candidate",state.term);
+        }catch(e){
+            log.fatal(util.inspect(e));
+        }finally {
+            let end = (new Date()).getTime();
+            let delay = end-begin;
+            if(delay>150){
+                log.debug(state.id, `follower.verifyLeader handler delay:`, `${delay}ms`,"leaderId:", state.leaderId, "self:", state);
+            }
+            mutex.unlock();
+        }
+    });
+
+    var changeRole = async function(newRole,newTerm){
+        if (state.role != "follower") {
+            log.warn('follower cycle in changeRole', "self:",state);
+            return;
+        }
+
+        state.leaderId = 0;
+        state.term = newTerm;
+        state.role = newRole;
+        log.info(`follower->${newRole}`,"self:",state);
+        await topicChannel.unsubscribe(routerKeys);
+        switch(newRole){
+            case "candidate":
+                await runAsCandidate(topicChannel, manager);
+                break;
+            case "follower":
+                await runAsFollower(topicChannel, manager);
+                break;
+
+        }
+    }
+
+    var installSnapshot = async function (data) {
+        if (!isInstallSnapshot) {
+            isInstallSnapshot = true;
+            log.info("start installSnapshot last commitId:",state.commitId,"data.commitId:",data.commitId,"from:",data.id);
+            topicChannel.publish(`clusterManager.leader.${state.leaderId}`, {type: 'installSnapshot', data: state});
+            return new Promise((resolve) => {
+                let count = 0;
+                let checkIntvl = setInterval(() => {
```

Comment:
Wait 20*25 ms to let installSnapshot finish here?
I think you can use a callback so the promise can resolve as soon as installSnapshot finishes.
```
var isInstallSnapshot = false;
var installSnapshotCallback = null;

return new Promise((resolve) => {
   let snapshotTimer = setTimeout(() => {
     isInstallSnapshot = false;
     resolve(false);
   }, 500);
   installSnapshotCallback = () => {
     clearTimeout(snapshotTimer);
     resolve(true);
   };
});

// ...
// When you set isInstallSnapshot to false
if (installSnapshotCallback) {
  installSnapshotCallback();
}
```

Reply:
ok，i will change it

Reply:
> Wait 20*25 ms to let installSnapshot finish here? I think you can use a callback so the promise can resolve as soon as installSnapshot finishes.
> 
> ```
> var isInstallSnapshot = false;
> var installSnapshotCallback = null;
> 
> return new Promise((resolve) => {
>    let snapshotTimer = setTimeout(() => {
>      isInstallSnapshot = false;
>      resolve(false);
>    }, 500);
>    installSnapshotCallback = () => {
>      clearTimeout(snapshotTimer);
>      resolve(true);
>    };
> });
> 
> // ...
> // When you set isInstallSnapshot to false
> if (installSnapshotCallback) {
>   installSnapshotCallback();
> }
> ```

ok，i will chang this func